### PR TITLE
Fixed issue #23638

### LINKED ISSUE #23638: Sample Files for Unsalable Products Redirect Instead of Download
### Preconditions (*)

1. 2.4-develop
2. Set Stores->Catalog->Inventory->Stock Options->Display Out of Stock Products=Yes

### Steps to reproduce (*)

1. Login to admin
2. Go to Catalog-->Products
3. Select 'Downloadable product' from 'Add product' button dropdown
4. Fill all required inputs 
5. Set product status to 'Out to stock'
6. Go to 'Downloadable information' tab
7. Press 'Add link' button under a Samples 
8. Add any test file
9. Save a product
10. Go to the product page on the frontend
11. Press on link under 'Samples'

### Expected result (*)

1. By default, the sample file should be downloaded. Products that are not currently for sale should still allow customers to download samples. 

### Actual result (*)

1. New tab is opened. No download is started and no explanation is given.

### PR DESCRIPTION
**Pull request by geet07**
### Description (*)

Sample Files for Unsalable Products Redirect Instead of Download

### Fixed Issues (if relevant)

1. magento/magento2#<#23638>: Sample Files for Unsalable Products Redirect Instead of Download

### Manual testing scenarios (*)

1. Create a product with a downloadable sample file.
2. Configure the product/store to make the product unsalable, yet still visible on the frontend.
3. On the frontend, find the product and click on the downloadable sample link.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by geet07** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @geet07. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @geet07, here is your [new Magento instance](https://pr-24705.instances.magento-community.engineering).
Admin access: [https://pr-24705.instances.magento-community.engineering/admin](https://pr-24705.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @geet07,
Thank you for your contribution!
Your changes looks good for me.

Would be great to cover your changes with MFTF test. Will you be able to do that?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @geet07,
I just double checked - this check was added in following commit https://github.com/magento/magento2/commit/484c8530afe0a2003d9f52517f64131ae04ac722
And was covered with MFTF test
https://github.com/magento/magento2/blob/e93feb02398f158f79fefb14de95cf93e3ec0786/app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml

Not sure why this test case didn't failed Functional tests, but it suppose to.

This test case describes that "Samples of Downloadable Products are not accessible, if product is disabled", so this check should be replaced to check if product is enabled.

Could you adjust your code?

**Comment by geet07** [comment]
> s covered with MFTF test

Hi @ihor-sviziev,

Thank you for your message.

I have confusion if the product is disabled then it is not showing in frontend. So there is why need to check sample with disabled products. And I have not much more idea for MFTF test cases. Can you please help me. Where I need to update code, in test cases or in functional code

**Comment by ihor-sviziev** [comment]
Hi @geet07,

I just asked to clarify actual steps to peproduce in https://github.com/magento/magento2/issues/23638#issuecomment-535410294

Let's wait for answer

**Comment by ihor-sviziev** [comment]
Hi @geet07,
Steps to reproduce were updated. Please see https://github.com/magento/magento2/issues/23638#issuecomment-535427010

**Comment by geet07** [comment]
> o reproduce were updated. Please see #23

Hi @ihor-sviziev,

Yes, I have seen the message and I have already done accordingly. Can you please share, what step I need to do

**Comment by ihor-sviziev** [comment]
@geet07 now we have regression - when product status is "Disabled" - it should NOT be available for download

**Comment by geet07** [comment]
> ct status is "Disabled"

ok, I have also tried to check this case with disabled products, but the disabled product is not showing in frontend. So this case not generated with disabled products and have tested with out of stock products and it's working fine.

**Comment by geet07** [comment]
@ihor-sviziev  ok, I have also tried to check this case with disabled products, but the disabled product is not showing in frontend. So this case not generated with disabled products and have tested with out of stock products and it's working fine.

**Comment by ihor-sviziev** [comment]
Hi @geet07,

I just checked on test instance - when product is disabled - [product link](https://pr-24705.instances.magento-community.engineering/test-download.html) returns 404 page ✔, but [link from "sample"](https://pr-24705.instances.magento-community.engineering/downloadable/download/sample/sample_id/1/) still works, but it shouldn't ❌.

If product is out of stock - Sample **should** be available.
If product is disabled - Sample **should NOT** be available.

**Comment by geet07** [comment]
@ihor-sviziev,

I will do it.

**Comment by ihor-sviziev** [comment]
@geet07, I just wanted to fix existing MFTF test that currently contains incorrect check, but I wasn't able to commit in your branch. So please apply also following change.
https://github.com/geeta07/magento2/compare/fix-23638...ihor-sviziev:patch-1

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Looks like following should be correct:
```suggestion
        <seeCurrentUrlEquals url="http://{$hostname}/" stepKey="seeRedirectToHomePage"/>
```

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Looks like following should be correct:
```suggestion
        <seeCurrentUrlEquals url="http://{$hostname}/" stepKey="seeHomePage"/>
```

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Controller/Download/Sample.php]: Please keep `SalabilityChecker ` parameter as it was before and add new argument. It's required due to [Backward Compatibility policy](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Controller/Download/Sample.php]: 1. Would be great to not depend on non-interface method. I think it will be enough to check `(int)$product->getStatus() === \Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED`.


2. Questions: What behavior will we have if:
- product isn't associated to this website?
- product is "disabled"?
- product visibility set to "Not visible individually"?

AFAIK In all these 3 cases product will not be available on frontend -> downaload sample should not be available either. 

Would be great do cover all these cases with MFTF tests. Will you be able to do that?

3. If we weren't able to load sample - we'll have NoSuchEntityException because `$sample->getProductId()` will return null

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Controller/Download/Sample.php:48]: Could you also add `\Magento\Downloadable\Model\SampleFactory` in constructor and create `$sample` object from it, not via ObjectManager?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (geet07) [reply app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Done this changes but getting error undefine $hostname

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (geet07) [reply app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Done

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (geet07) [reply app/code/Magento/Downloadable/Controller/Download/Sample.php]: I have changed the functionality, Please check accordingly.

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (geet07) [reply app/code/Magento/Downloadable/Controller/Download/Sample.php:48]: Done

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (geet07) [reply app/code/Magento/Downloadable/Controller/Download/Sample.php]: I have changed the functionality, Please check accordingly.

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Previous version didn’t worked, let’s try in this way
```suggestion
        <seeCurrentUrlEquals url="{{_ENV.MAGENTO_ BASE_URL}}" stepKey="seeRedirectToHomePage"/>
```

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (ihor-sviziev) [inline app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]:
```suggestion
        <seeCurrentUrlEquals url="{{_ENV.MAGENTO_ BASE_URL}}" stepKey="seeHomePage"/>
```

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/SalabilityChecker.php:74]: Unfortunately adding new methods to existing class - backward incompatible change. please move this logic inside constructor

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (ihor-sviziev) [inline app/code/Magento/Catalog/Model/Product/SalabilityChecker.php]: There actually might be exception, please wrap it to try-catch block and return false if we weren’t able to find such product

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @geet07,
Could you review my comments? 
Also could you enable option to allow edits from maintainers, I'll be able to adjust tests for you https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (sidolov) [inline app/code/Magento/Catalog/Model/Product/SalabilityChecker.php]: Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/)

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (geet07) [reply app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: I have updated this still getting something error.

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (geet07) [reply app/code/Magento/Catalog/Model/Product/SalabilityChecker.php:74]: I have updated code according to backward compatibility.

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (geet07) [reply app/code/Magento/Catalog/Model/Product/SalabilityChecker.php]: Done

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (geet07) [reply app/code/Magento/Catalog/Model/Product/SalabilityChecker.php]: Thanks for sharing this link, I have updated my code accordingly.

File: app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml
Comment (ihor-sviziev) [reply app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Please remove space there`{{_ENV.MAGENTO_BASE_URL}}`

Comment (geet07) [reply app/code/Magento/Downloadable/Test/Mftf/Test/VerifyDisableDownloadableProductSamplesAreNotAccessibleTest.xml]: Done

File: app/code/Magento/Downloadable/Controller/Download/Sample.php
Comment (ihor-sviziev) [reply app/code/Magento/Downloadable/Controller/Download/Sample.php:48]: Sorry, still can’t find these changes, could you double check it?

File: app/code/Magento/Catalog/Model/Product/SalabilityChecker.php
Comment (ihor-sviziev) [reply app/code/Magento/Catalog/Model/Product/SalabilityChecker.php:74]: Sorry, still can’t see your changes, please double check

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@geet07 I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**
